File: scripts/yyb_account_guard.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# name: yyb_account_guard
# cron: 15 16 * * *
"""YYB 多账号公共状态缓存。

业务脚本只需要在遍历 YYB_SERVER 前调用 ``filter_accounts``，在捕获业务
错误时调用 ``mark_from_error``。缓存是共享的，默认放在青龙持久化目录。
"""

from __future__ import annotations

import json
import os
import re
import tempfile
import time
from pathlib import Path
from typing import Any, Callable, Iterable, TypeVar


T = TypeVar("T")

STATUS_FILE_ENV = "YYB_ACCOUNT_STATUS_FILE"
DEFAULT_STATUS_FILE = "/ql/data/config/yyb_account_status.json"
STATUS_VERSION = 1

# 明确表示业务账号未完成初始化的文本。只在命中这些词时跳过。
UNBOUND_PATTERNS = (
    r"未授权手机号",
    r"手机号未授权",
    r"未完成手机号授权",
    r"请先.{0,16}授权手机号",
    r"尚未注册.{0,16}(?:会员|小程序)",
    r"未注册(?:会员|小程序)",
    r"未绑定(?:此小程序|手机号|会员|账号)",
    r"尚未绑定",
    r"未登录.{0,12}(?:会员|账号)",
    r"微信授权成功.{0,20}未登录",
)

# 这些词即使和“登录失败”同时出现，也不能标记为未注册。
TEMPORARY_PATTERNS = (
    r"timeout",
    r"timed out",
    r"超时",
    r"HTTP\s*[45]\d\d",
    r"\b(?:502|503|504)\b",
    r"服务端异常",
    r"系统繁忙",
    r"活动太火爆",
    r"风控",
    r"third login fds limit",
    r"登录过期",
    r"token.{0,8}(?:失效|过期)",
)

_UNBOUND_RE = re.compile("|".join(UNBOUND_PATTERNS), re.IGNORECASE)
_TEMPORARY_RE = re.compile("|".join(TEMPORARY_PATTERNS), re.IGNORECASE)

TTL_SECONDS = {
    "unbound": 24 * 60 * 60,
    "unregistered": 24 * 60 * 60,
    "code_unavailable": 30 * 60,
    "temporary_error": 10 * 60,
    "ready": 6 * 60 * 60,
    "unknown": 15 * 60,
}


def status_file() -> Path:
    configured = os.getenv(STATUS_FILE_ENV, "").strip()
    if configured:
        return Path(configured).expanduser()
    default = Path(DEFAULT_STATUS_FILE)
    if default.parent.exists() or os.getenv("QL_DATA_DIR"):
        return default
    # 本地开发/测试时不污染仓库，跟随脚本目录保存。
    return Path(__file__).resolve().parent / "yyb_account_status.json"


def account_ref(value: Any) -> str:
    """从 YYB_SERVER 行、对象或裸 ref 中取得稳定账号标识。"""
    if isinstance(value, dict):
        value = value.get("ref") or value.get("id") or value.get("openid") or ""
    text = str(value or "").strip()
    if "@" in text:
        text = text.rsplit("@", 1)[1].strip()
    return text


def _read() -> dict[str, Any]:
    path = status_file()
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(payload, dict) and isinstance(payload.get("accounts"), dict):
            return payload
    except (OSError, ValueError, TypeError):
        pass
    return {"version": STATUS_VERSION, "updated_at": 0, "accounts": {}}


def _write(payload: dict[str, Any]) -> None:
    path = status_file()
    path.parent.mkdir(parents=True, exist_ok=True)
    payload["version"] = STATUS_VERSION
    payload["updated_at"] = int(time.time())
    fd, temp_name = tempfile.mkstemp(prefix=".yyb-status-", suffix=".tmp", dir=str(path.parent))
    try:
        with os.fdopen(fd, "w", encoding="utf-8", newline="\n") as handle:
            json.dump(payload, handle, ensure_ascii=False, indent=2)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temp_name, path)
    finally:
        try:
            os.unlink(temp_name)
        except FileNotFoundError:
            pass


def classify_error(message: Any) -> tuple[str, str] | None:
    """返回可缓存的状态；临时故障返回 ``temporary_error``，其余返回 None。"""
    text = str(message or "").replace("\n", " ").strip()
    if not text:
        return None
    if _TEMPORARY_RE.search(text):
        return "temporary_error", text[:240]
    if _UNBOUND_RE.search(text):
        lowered = text.lower()
        status = "unregistered" if ("注册" in text or "会员" in text) else "unbound"
        return status, text[:240]
    return None


def _storage_key(ref: Any, app_id: str = "") -> str:
    key = account_ref(ref)
    return f"{key}::{app_id}" if key and app_id else key


def get_status(ref: Any, *, app_id: str = "") -> dict[str, Any] | None:
    key = account_ref(ref)
    if not key:
        return None
    accounts = _read().get("accounts", {})
    if app_id:
        return accounts.get(_storage_key(key, app_id))
    return accounts.get(key)


def should_skip(ref: Any, *, app_id: str = "", now: float | None = None) -> tuple[bool, str]:
    key = account_ref(ref)
    if not key or os.getenv("YYB_GUARD_BYPASS", "").strip() == "1":
        return False, ""
    item = get_status(key, app_id=app_id)
    if not isinstance(item, dict):
        return False, ""
    status = str(item.get("status", "unknown"))
    if status == "disabled":
        return True, str(item.get("reason") or "账号已被公共缓存停用")
    checked = float(item.get("checked_at") or 0)
    ttl = int(TTL_SECONDS.get(status, TTL_SECONDS["unknown"]))
    current = time.time() if now is None else now
    if checked and current - checked < ttl and status in {"unbound", "unregistered", "code_unavailable", "temporary_error"}:
        return True, str(item.get("reason") or status)
    return False, ""


def set_status(ref: Any, status: str, reason: str = "", *, app_id: str = "") -> None:
    ref_key = account_ref(ref)
    key = _storage_key(ref_key, app_id)
    if not key:
        return
    payload = _read()
    accounts = payload.setdefault("accounts", {})
    accounts[key] = {
        "ref": ref_key,
        "status": status,
        "reason": str(reason or "")[:240],
        "app_id": str(app_id or ""),
        "checked_at": int(time.time()),
    }
    _write(payload)


def mark_from_error(ref: Any, message: Any, *, app_id: str = "") -> str | None:
    classified = classify_error(message)
    if not classified:
        return None
    status, reason = classified
    set_status(ref, status, reason, app_id=app_id)
    return status


def mark_ready(ref: Any, *, app_id: str = "") -> None:
    set_status(ref, "ready", "", app_id=app_id)


def update_from_result(ref: Any, result: Any, *, app_id: str = "") -> str | None:
    """根据脚本统一结果回写状态。

    只有明确的成功或可分类错误才写缓存；未知异常不改变已有状态，避免
    把接口字段变化误判成“账号未授权”。
    """
    if not isinstance(result, dict):
        return None
    if result.get("success") is True:
        mark_ready(ref, app_id=app_id)
        return "ready"
    message = result.get("error") or result.get("message") or result.get("msg") or ""
    return mark_from_error(ref, message, app_id=app_id)


def filter_accounts(
    values: Iterable[T],
    ref_getter: Callable[[T], Any] | None = None,
    *,
    app_id: str = "",
    log: Callable[[str], None] | None = print,
) -> list[T]:
    """过滤冷却中的账号，并保持原顺序。"""
    getter = ref_getter or (lambda value: value)
    kept: list[T] = []
    for value in values:
        ref = account_ref(getter(value))
        skip, reason = should_skip(ref, app_id=app_id)
        if skip:
            if log:
                log(f"账号 {ref} 已由 YYB 公共缓存跳过：{reason}")
            continue
        kept.append(value)
    return kept


def prune(*, refs: Iterable[Any] | None = None) -> dict[str, Any]:
    """清理不再存在于 YYB_SERVER 的账号状态，返回汇总。"""
    payload = _read()
    accounts = payload.setdefault("accounts", {})
    allowed = {account_ref(ref) for ref in refs or () if account_ref(ref)}
    removed: list[str] = []
    if allowed:
        for key in list(accounts):
            base_key = key.split("::", 1)[0]
            if base_key not in allowed:
                removed.append(key)
                accounts.pop(key, None)
    _write(payload)
    return {"total": len(accounts), "removed": removed, "file": str(status_file())}


__all__ = [
    "account_ref",
    "classify_error",
    "filter_accounts",
    "get_status",
    "mark_from_error",
    "mark_ready",
    "update_from_result",
    "prune",
    "set_status",
    "should_skip",
    "status_file",
]

File: scripts/yyb-account-guard.js
// name: yyb-account-guard
"use strict";

// Node 版公共缓存；业务脚本在遍历 YYB_SERVER 前调用 filterAccounts。
const fs = require("fs");
const os = require("os");
const path = require("path");

const STATUS_FILE = process.env.YYB_ACCOUNT_STATUS_FILE ||
  (fs.existsSync("/ql/data/config") ? "/ql/data/config/yyb_account_status.json" :
    path.join(__dirname, "yyb_account_status.json"));
const TTL = {
  unbound: 24 * 60 * 60,
  unregistered: 24 * 60 * 60,
  code_unavailable: 30 * 60,
  temporary_error: 10 * 60,
  ready: 6 * 60 * 60,
  unknown: 15 * 60,
};
const UNBOUND = /未授权手机号|手机号未授权|未完成手机号授权|请先.{0,16}授权手机号|尚未注册.{0,16}(会员|小程序)|未注册(会员|小程序)|未绑定(此小程序|手机号|会员|账号)|尚未绑定|未登录.{0,12}(会员|账号)|微信授权成功.{0,20}未登录/i;
const TEMPORARY = /timeout|timed out|超时|HTTP\s*[45]\d\d|\b(502|503|504)\b|服务端异常|系统繁忙|活动太火爆|风控|third login fds limit|登录过期|token.{0,8}(失效|过期)/i;

function refOf(value) {
  if (value && typeof value === "object") value = value.ref || value.id || value.openid || "";
  const text = String(value || "").trim();
  return text.includes("@") ? text.slice(text.lastIndexOf("@") + 1).trim() : text;
}
function read() {
  try {
    const data = JSON.parse(fs.readFileSync(STATUS_FILE, "utf8"));
    if (data && data.accounts && typeof data.accounts === "object") return data;
  } catch (_) {}
  return { version: 1, updated_at: 0, accounts: {} };
}
function write(data) {
  fs.mkdirSync(path.dirname(STATUS_FILE), { recursive: true });
  const temp = path.join(path.dirname(STATUS_FILE), `.${path.basename(STATUS_FILE)}.${process.pid}.${Date.now()}.tmp`);
  data.version = 1;
  data.updated_at = Math.floor(Date.now() / 1000);
  fs.writeFileSync(temp, JSON.stringify(data, null, 2) + os.EOL, "utf8");
  fs.renameSync(temp, STATUS_FILE);
}
function classifyError(message) {
  const text = String(message || "").replace(/[\r\n]+/g, " ").trim();
  if (!text) return null;
  if (TEMPORARY.test(text)) return { status: "temporary_error", reason: text.slice(0, 240) };
  if (UNBOUND.test(text)) return { status: /注册|会员/.test(text) ? "unregistered" : "unbound", reason: text.slice(0, 240) };
  return null;
}
function setStatus(ref, status, reason = "", appId = "") {
  const key = refOf(ref);
  if (!key) return;
  const data = read();
  const storageKey = appId ? `${key}::${appId}` : key;
  data.accounts[storageKey] = { ref: key, status, reason: String(reason || "").slice(0, 240), app_id: appId || "", checked_at: Math.floor(Date.now() / 1000) };
  write(data);
}
function markFromError(ref, message, appId = "") {
  const result = classifyError(message);
  if (!result) return null;
  setStatus(ref, result.status, result.reason, appId);
  return result.status;
}
function shouldSkip(ref, appId = "") {
  const key = refOf(ref);
  if (!key || process.env.YYB_GUARD_BYPASS === "1") return { skip: false, reason: "" };
  const accounts = read().accounts;
  const item = accounts[appId ? `${key}::${appId}` : key];
  if (!item) return { skip: false, reason: "" };
  if (item.status === "disabled") return { skip: true, reason: item.reason || "账号已停用" };
  const ttl = TTL[item.status] || TTL.unknown;
  const checked = Number(item.checked_at || 0);
  if (checked && Date.now() / 1000 - checked < ttl && ["unbound", "unregistered", "code_unavailable", "temporary_error"].includes(item.status)) {
    return { skip: true, reason: item.reason || item.status };
  }
  return { skip: false, reason: "" };
}
function filterAccounts(values, refGetter = value => value, options = {}) {
  const appId = typeof options === "string" ? options : options.appId || "";
  const log = typeof options === "object" ? (options.log || console.log) : console.log;
  return values.filter(value => {
    const ref = refOf(refGetter(value));
    const result = shouldSkip(ref, appId);
    if (result.skip && log) log(`账号 ${ref} 已由 YYB 公共缓存跳过：${result.reason}`);
    return !result.skip;
  });
}

module.exports = { STATUS_FILE, classifyError, filterAccounts, markFromError, refOf, setStatus, shouldSkip };
